velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -13,89 +13,214 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 #include "velox/exec/Aggregate.h"
 #include "velox/expression/FunctionSignature.h"
 #include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/functions/prestosql/aggregates/SingleValueAccumulator.h"
 #include "velox/vector/ComplexVector.h"
 #include "velox/vector/DecodedVector.h"
 #include "velox/vector/FlatVector.h"
-
 namespace facebook::velox::aggregate {
 
 namespace {
 
-// MinMaxByAggregate is the base class for min_by and max_by functions. These
-// functions return the value of X associated with the minimum/maximum value of
-// Y over all input values. Partial aggregation produces a pair of X and min/max
-// Y. Final aggregation takes a pair of X and min/max Y and returns X. T is the
-// type of X and U is the type of Y.
+template <typename T>
+constexpr bool isNumeric() {
+  return std::is_same_v<T, int8_t> || std::is_same_v<T, int16_t> ||
+      std::is_same_v<T, int32_t> || std::is_same_v<T, int64_t> ||
+      std::is_same_v<T, float> || std::is_same_v<T, double>;
+}
+
+template <typename T, typename TAccumulator>
+void extract(
+    TAccumulator* accumulator,
+    const VectorPtr& vector,
+    vector_size_t index,
+    T* rawValues) {
+  if constexpr (isNumeric<T>()) {
+    rawValues[index] = *accumulator;
+  } else {
+    VELOX_DCHECK(accumulator->hasValue());
```

Comment:
This check is redundant since SingleValueAccumulator::read already has it:

```
void SingleValueAccumulator::read(const VectorPtr& vector, vector_size_t index)
    const {
  VELOX_CHECK(begin_);
...
bool SingleValueAccumulator::hasValue() const {
  return begin_ != nullptr;
}
```